### PR TITLE
Majority calculation for assignments

### PR DESCRIPTION
**Pull request by matakuka**
Majority calculation adapted for assignments, too.
What is still missing: relative majorities (i.e., "first two candidates"), but I think this may better be another PR.

### CODE REVIEW

File: openslides/assignments/static/templates/assignments/assignment-detail.html [hunk 1]
```diff
@@ -126,120 +126,135 @@ <h3 translate>Election result</h3>
...
-              </span>
-              <a ui-sref="users.user.detail({id: option.candidate.id})">{{ option.candidate.get_full_name() }}</a>
+          <!-- Settings for majority calculations -->
+          <div os-perms="assignments.can_manage" ng-hide="config('assignments_poll_default_majority_method') == 'disabled'" ng-cloak>
+            <div class="input-group">
+              <span><translate>Required majority</translate>: </span>
+              <select ng-model="method"
+                ng-change="recalculateMajorities(method)" ng-options="option.value as option.display_name for option in methodChoices" />
+            </div>
+            {{ method }}
```
Comment (emanuelschuetze) [inline openslides/assignments/static/templates/assignments/assignment-detail.html]: Please remove this debug line.

File: openslides/assignments/static/templates/assignments/assignment-detail.html [hunk 2]
```diff
@@ -126,120 +126,135 @@ <h3 translate>Election result</h3>
...
-                    title="{{ 'is not elected' | translate }}"></i>
-                &nbsp;
-              </span>
-              <a ui-sref="users.user.detail({id: option.candidate.id})">{{ option.candidate.get_full_name() }}</a>
+          <!-- Settings for majority calculations -->
+          <div os-perms="assignments.can_manage" ng-hide="config('assignments_poll_default_majority_method') == 'disabled'" ng-cloak>
+            <div class="input-group">
+              <span><translate>Required majority</translate>: </span>
+              <select ng-model="method"
+                ng-change="recalculateMajorities(method)" ng-options="option.value as option.display_name for option in methodChoices" />
```
Comment (emanuelschuetze) [inline openslides/assignments/static/templates/assignments/assignment-detail.html]: use `... option.display_name | translate for ...`  to translate display name.

File: openslides/assignments/static/templates/assignments/assignment-detail.html [hunk 3]
```diff
@@ -126,120 +126,135 @@ <h3 translate>Election result</h3>
...
-                  ng-click="markElected(option.candidate_id, option.is_elected)"
-                    title="{{ 'is elected' | translate }}"></i>
-                <i ng-if="!option.is_elected" class="fa fa-square-o"
-                  ng-click="markElected(option.candidate_id, option.is_elected)"
-                    title="{{ 'is not elected' | translate }}"></i>
-                &nbsp;
-              </span>
-              <a ui-sref="users.user.detail({id: option.candidate.id})">{{ option.candidate.get_full_name() }}</a>
+          <!-- Settings for majority calculations -->
+          <div os-perms="assignments.can_manage" ng-hide="config('assignments_poll_default_majority_method') == 'disabled'" ng-cloak>
```
Comment (emanuelschuetze) [inline openslides/assignments/static/templates/assignments/assignment-detail.html]: please hide this if poll has no votes `ng-hide=" ... && !poll.has_votes"`

File: openslides/poll/static/js/poll/majority.js [hunk 4]
```diff
@@ -76,8 +127,10 @@ angular.module('OpenSlidesApp.poll.majority', [])
                         break;
                     // case 'DISABLED': result remains undefined
                 }
+                console.log(result);
+                console.log("Result");
```
Comment (emanuelschuetze) [inline openslides/poll/static/js/poll/majority.js]: Please remove this lines.

File: openslides/poll/static/js/poll/majority.js [hunk 5]
```diff
@@ -38,29 +38,80 @@ angular.module('OpenSlidesApp.poll.majority', [])
...
+                var error = false;
+                if (poll.options) {
+                    _.forEach(poll.options, function(option) {
+                        _.forEach(option.votes, function(vote) {
+                            if (vote.value == 'Yes' || vote.value == 'Votes'){
+                                if (vote.weight >= 0){
+                                    yes = yes + vote.weight;
+                                } else {
+                                    error = true;
+                                    //TODO abort searching here
```
Comment (emanuelschuetze) [inline openslides/poll/static/js/poll/majority.js]: ???

File: openslides/assignments/static/templates/assignments/assignment-detail.html [hunk 6]
```diff
@@ -126,120 +126,135 @@ <h3 translate>Election result</h3>
...
+              <td>
+                <span os-perms="assignments.can_manage">
+                  <i ng-if="option.is_elected" class="fa fa-check-square-o"
+                    ng-click="markElected(option.candidate_id, option.is_elected)"
+                      title="{{ 'is elected' | translate }}"></i>
+                  <i ng-if="!option.is_elected" class="fa fa-square-o"
+                    ng-click="markElected(option.candidate_id, option.is_elected)"
+                      title="{{ 'is not elected' | translate }}"></i>
+                  &nbsp;
+  <!--               </span> -->
```
Comment (emanuelschuetze) [inline openslides/assignments/static/templates/assignments/assignment-detail.html]: brocken comment

Comment (emanuelschuetze) [inline openslides/assignments/static/templates/assignments/assignment-detail.html]: You should close this span tag...

**Comment by emanuelschuetze** [comment]
@matakuka Please rebase and resolve the conflicts.

**Comment by matakuka** [comment]
I rebased it on top of master (I did not squish my commits yet) and integrated @emanuelschuetze s requests

**Comment by emanuelschuetze** [comment]
Ok.

@normanjaeckel Please review

**Comment by normanjaeckel** [comment]
Caluclation seems broken:

bildschirmfoto vom 2016-11-02 14 51 23

The problem is the sumofvotes argument I think.

File: openslides/assignments/static/js/assignments/site.js [hunk 7]
```diff
@@ -229,6 +230,62 @@ angular.module('OpenSlidesApp.assignments.site', [
...
+                base: $scope.config('assignments_poll_100_percent_base')
+            };
+        }
+
+        // Fetch users choices from cache.
+        $scope.isMajorityCalculation = AssignmentPollDetailCtrlCache[$scope.poll.id].isMajorityCalculation;
+        $scope.isMajorityDetails = AssignmentPollDetailCtrlCache[$scope.poll.id].isMajorityDetails;
+        $scope.method = AssignmentPollDetailCtrlCache[$scope.poll.id].method;
+        $scope.base = AssignmentPollDetailCtrlCache[$scope.poll.id].base;
+        $scope.totalyesvotes = MajorityCalculation.options_yes_sum($scope.poll);
```
Comment (normanjaeckel) [inline openslides/assignments/static/js/assignments/site.js]: Insert a newline and a new comment here. This is a new codeblock.

File: openslides/assignments/static/js/assignments/site.js [hunk 8]
```diff
@@ -836,6 +893,13 @@ angular.module('OpenSlidesApp.assignments.site', [
         gettext('Number of all participants');
         gettext('Use the following custom number');
         gettext('Custom number of ballot papers');
+        gettext('Required majority');
+        gettext('Default method to check whether a candidate has reached the required majority.');
+        gettext('Relative majority');
```
Comment (normanjaeckel) [inline openslides/assignments/static/js/assignments/site.js]: This string does not exist in OpenSlides, does is?

**Comment by matakuka** [comment]
Calculation is correct according to the formula we defined, assuming that the "base" was Yes/no(/abstain). As we cannpot be sure that y+n+a for all candidates is the same, we preferred the sum of total affirmative votes (see [config_variables.py: 39-41](https://github.com/OpenSlides/OpenSlides/blob/master/openslides/assignments/config_variables.py#L39-L41)

I agree that the screenshot is definitely wrong. I think this is rather an issue in the percentage calculation not implementing our rather complicated condition.

Which should I change?

**Comment by normanjaeckel** [comment]
If we have a YNA election, the sum of votes is the sum of YNA. For candidate dgdfg it is 6+7+8=21. The calculation says that the quorum is reached with 1 vote but the simple majority quorum of 21 is 11, not 5. I think your algorithm adds 3+6 = 9 in the example above which is wrong.

**Comment by normanjaeckel** [comment]
If we have a YNA election, the sum of votes is the sum of YNA. For candidate dgdfg it is 6+7+8=21. The calculation says that the quorum is reached with 1 vote but the simple majority quorum of 21 is 11, not 5. I think your algorithm adds 3+6 = 9 in the example above which is wrong.

**Comment by normanjaeckel** [comment]
If we have a YNA election, the sum of votes is the sum of YNA. For candidate dgdfg it is 6+7+8=21. The calculation says that the quorum is reached with 1 vote but the simple majority quorum of 21 is 11, not 5. I think your algorithm adds 3+6 = 9 in the example above which is wrong.

**Comment by normanjaeckel** [comment]
Please also care that the number of open posts and the number of candidates does not necessarily determine the election method. You can setup a special config for this so you can not relay of the number of posts and candidates in this context.

**Comment by matakuka** [comment]
done @normanjaeckel

**Comment by normanjaeckel** [comment]
Quorum check is missing if config is set to YNA or YN and we have the simple election method (one vote per candidate).

bildschirmfoto vom 2016-11-08 15 53 35

**Comment by matakuka** [comment]
I still have a minor problem and don't know how to circumvent it:
If the election method is changed after an election (e.g. from "always Y/N/A to one option per candidate), older calculations will be wrong. I cannot reliably see which method a poll used, only which method is *currently* active.

**Comment by normanjaeckel** [comment]
The election method is save in the poll. After a new poll is created its election method can not change.

**Comment by matakuka** [comment]
@normanjaeckel: another attempt at getting it right. Sorry for all the confusion and delay.

**Comment by normanjaeckel** [comment]
@matakuka I pushed force to your branch. So please be careful here.

I refactored the percents and majority checks completely. @emanuelschuetze and @matakuka Please review and test carefully.

One new thing: If we have percents for Yes and No only then I disabled percents and majorities if abstain is set to -1. I think you can not have abstain -1 (wording "majority") but nevertheless expect some percents and majorities for yes and no. Of course it is possible to set abstain to -2 or 0 or anything else. What do you think?

I also added percent bar for 0 %. Without the bar it looks somehow broken.

Slides are not checked. Please also check them.

**Comment by emanuelschuetze** [comment]
Tested:
* please simplify the quorum message: 
  "Quorum reached. Only X votes required." -> "Quorum (X) reached."
  "Quorum not reached. X votes required." -> "Quorum (X) not reached."
* assignment detail, vote result table: 
1. rename column name "Quorum reached" to "Quorum"
2. hide quorum column if majority check is disabled
3. percentage/quorum not updated if 100% base in assignment config is changed (e.g. create ballot with 1 candidate with YNA method. Change 100% base in 2nd tab from YNA -> YN. Percent not updated. Reload shows correct percent values.)

**Comment by normanjaeckel** [comment]
Quorum message changed. Column renamed and hidden in case of method === disabled. Show method dropdown for motions even if config is set to disabled. Opened issues #2694 and  #2695.

**Comment by emanuelschuetze** [comment]
@normanjaeckel Please rebase and merge.

**Pull request merged**